rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,653 @@
+
+When writing this section be mindful of the following [repo guidelines](https://github.com/bevyengine/rfcs):
+
+- **RFCs should be scoped:** Try to avoid creating RFCs for huge design spaces that span many features. Try to pick a specific feature slice and describe it in as much detail as possible. Feel free to create multiple RFCs if you need multiple features.
+- **RFCs should avoid ambiguity:** Two developers implementing the same RFC should come up with nearly identical implementations.
+- **RFCs should be "implementable":** Merged RFCs should only depend on features from other merged RFCs and existing Bevy features. It is ok to create multiple dependent RFCs, but they should either be merged at the same time or have a clear merge order that ensures the "implementable" rule is respected. -->
+
+### Names
+
+Being data-heavy, and forfeiting the structural access of Rust types, the data model makes extensive use of strings to reference items. In order to make those operations efficient, we define a `Name` type to represent those strings. The implementation of `Name` is outside of the scope of this RFC and left as an optimization, and should be considered as equivalent to a `String` wrapper:
+
+```rust
+struct Name(pub String);
+// -or-
+type Name = String;
+```
+
+The intent is to explore some optimization strategies like string interning for fast comparison. Those optimizations are left out of the scope of this RFC.
+
+### Built-in types
+
+The data model supports a number of built-in types, corresponding to common Rust built-in types.
+
+```rust
+enum SignedIntegralType {
+    Int8, Int16, Int32, Int64, Int128,
+}
+
+enum UnsignedIntegralType {
+    Uint8, Uint16, Uint32, Uint64, Uint128,
+}
+
+enum FloatingType {
+    Float32, Float64,
+}
+
+enum SimpleType {
+    Bool,
+    Int8, Int16, Int32, Int64, Int128,
+    Uint8, Uint16, Uint32, Uint64, Uint128,
+    Float32, Float64,
+}
+
+enum BuiltInType {
+    Bool,
+    Int8, Int16, Int32, Int64, Int128,
+    Uint8, Uint16, Uint32, Uint64, Uint128,
+    Float32, Float64,
+    Name,
+    Enum(UnsignedIntegralType), // C-style enum, not Rust-style
+    Array(AnyType),
+    Dict(AnyType), // key type is Name
+    ObjectRef(Option<Entity>),
+}
+```
+
+_Note_: We explicitly handle all integral and floating-point bit variants to ensure tight value packing and avoid wasting space both in structs (`GameObject`; see below) and collections (arrays, _etc._).
+
+`ObjectRef` is a reference to any object, custom or built-in. This allows referencing components built-in inside Bevy or the Editor (TBD depending on design of that part). The reference needs to be valid (non-null) when the data gets processed for baking, but can temporarily (including while serialized to disk and later reloaded) be left invalid (`None`) for editing flexibility. Nullable references require a `NullableRef` component to be added to mark the reference as valid even if null.

```

Comment:
I think you're maybe mixing what Rust allows you to do in a program vs. what the data model requests for data to be valid.

If my Game contains a struct `S` with an `Entity` field, I need to assign a valid entity to it, as per Rust rules.

```rust
// Game process
struct S {
  entity: Entity, // must be valid
}
```

That means that when the Game deserialize an instance, that instance needs to contain a valid value.

Now on the Editor side in the data model, that field is represented by `Option<Entity>`. The reason is that I'm allowed to instantiate a dynamic editor object of `S` without an entity assigned yet, and do other things, save my projects, _etc._ and that's fine.

```rust
// Editor process
let entity_type = gameTypeFromName.get("Entity"); // equivalent to "Option<Entity>"
let s_type = GameType::new().name("S").add_field("entity", entity_type);
let s_instance = GameObject::new(s_type);
s_instance.set_entity_field("entity", Some(myEntity)); // OK, valid value
s_instance.set_entity_field("entity", None); // Allowed, but makes the data temporarily invalid
```

The only thing is that my data is in an invalid state, so the Editor will refuse to bake it into Game-ready data, because the Game wouldn't be able to deserialize a null entity into `S::entity`. Once someone (a designer, an artist, whoever) fixed that field in the Editor to assign a valid entity, the data is valid and can be baked.

In contrast, I might have another field which in my Game is an `Option<Entity>`, which means I may or may not need a valid entity there. In the Editor, this is represented as an `Option<Entity>` with a `NullableRef` marker, which tells the Editor "yes that field can be `None` and that's valid, you're allowed to bake that data, the Game can handle it".